internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,156 @@
 	}
+	if b.current.buffer == nil {
+		b.current.buffer = make([]byte, 0, b.bufferSize)
+		b.current.offset.end = b.bufferSize
+	}
+	return nil
 }
 
-func (b *InMemoryWriteBuffer) WriteAt(data []byte, offset int64) error {
+// Allocates memory to flushedBuffer if it hasn't been allocated already.
+func (b *InMemoryWriteBuffer) ensureFlushedBuffer() error {
+	if b.bufferSize == 0 {
+		return fmt.Errorf(ZeroSizeBufferError)
+	}
+	if b.flushed.buffer == nil {
+		b.flushed.buffer = make([]byte, 0, b.bufferSize)
+	}
 	return nil
 }
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
+	dataSize := int64(len(content))
+	receivedContentEndOffset := offset + dataSize
+	if dataSize == 0 {
+		return nil
+	}
+
+	// Ensure b.currentBuffer != nil
+	if err := b.ensureCurrentBuffer(); err != nil {
+		return err
+	}
+
+	// If receivedContentStartOffset is not in the range of offsets stored in current
+	// buffer, trigger temp file flow for writes.
+	if offset < b.current.offset.start ||
+		offset > b.current.offset.end {
+		// TODO: finalise write and trigger temp file flow.
+		logger.Debugf("TODO: finalise write and trigger temp file flow.")
+		return fmt.Errorf(NonSequentialWriteError)
+	}
+
+	// If data can be completely written to the current buffer, write it.
+	if receivedContentEndOffset <= b.current.offset.end {
+		return b.copyDataToBuffer(offset, receivedContentEndOffset, content)
+	}
+	// Else, write the chunk of data that can be written to the current buffer block.
+	l := b.current.offset.end - offset
+	err := b.copyDataToBuffer(offset, b.current.offset.end, content[:l])
+	if err != nil {
+		return err
+	}
+	//TODO: get status of last write to GCS. If unsuccessful, trigger temp file flow.
+	logger.Debugf("TODO: get status of last write to GCS and if " +
+		"unsuccessful, trigger temp file flow.")
+	if err = b.advanceToNextChunk(); err != nil {
+		return err
+	}
+	// TODO: trigger async upload to GCS for flushedBuffer.
+	logger.Debugf("TODO: trigger async upload to GCS for flushedBuffer.")
+
+	// Write the remaining data to currentBuffer.
+	return b.copyDataToBuffer(b.current.offset.start, receivedContentEndOffset, content[l:])
+}
+
+// Copies received content to currentChunk if it has capacity.
+func (b *InMemoryWriteBuffer) copyDataToBuffer(contentStartOffset, contentEndOffset int64, content []byte) error {
+	dataSize := int64(len(content))
+	si := contentStartOffset % b.bufferSize
+	ei := si + dataSize
+	if ei > b.bufferSize {
+		return fmt.Errorf(NotEnoughSpaceInCurrentBuffer)
+	}
+
+	copy(b.current.buffer[si:ei], content)
+	b.updateFileSize(contentEndOffset)
+	return nil
+}
+
+func (b *InMemoryWriteBuffer) advanceToNextChunk() error {
+	// ensure b.flushed.buffer != nil
+	if err := b.ensureFlushedBuffer(); err != nil {
+		return err
+	}
+
+	// Move current buffer to flushed buffer.
+	b.flushed.buffer, b.current.buffer = b.current.buffer, b.flushed.buffer
+	b.flushed.offset.start = b.current.offset.start
```

Comment:
Couldn't you just instead do (i) swap b.flushed and b.current, and then (b) b.current.offset.advanceBy(2*bufferSize)? That'll be much simpler and easier to read.

Reply:
(ii) -> This will require additional if checks as in the first call to advanceToNextChunk, flushed buffer start and end offset will be 0 which unnecessarily complicates the logic.

I've made some changes to simplify this though, PTAL.